plugincontainer/container_runner.go

```diff
@@ -0,0 +1,230 @@
+package plugincontainer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"runtime"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/go-plugin/runner"
+	"github.com/hashicorp/go-secure-stdlib/plugincontainer/config"
+)
+
+var (
+	_ runner.Runner = (*ContainerRunner)(nil)
+
+	ErrUnsupportedOS = errors.New("plugincontainer currently only supports Linux")
+)
+
+const pluginSocketDir = "/tmp/go-plugin-container"
+
+// ContainerRunner implements the Executor interface by running a container.
+type ContainerRunner struct {
+	logger hclog.Logger
+
+	hostSocketDir string
+
+	containerConfig *container.Config
+	hostConfig      *container.HostConfig
+	networkConfig   *network.NetworkingConfig
+
+	dockerClient *client.Client
+	stdout       io.ReadCloser
+	stderr       io.ReadCloser
+
+	id string
+}
+
+// NewContainerRunner must be passed a cmd that hasn't yet been started.
+func NewContainerRunner(logger hclog.Logger, cmd *exec.Cmd, cfg *config.ContainerConfig, hostSocketDir string) (*ContainerRunner, error) {
+	if runtime.GOOS != "linux" {
+		return nil, ErrUnsupportedOS
+	}
+
+	client, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, err
+	}
+
+	// Container config.
+	containerConfig := &container.Config{
+		Env:             cmd.Env,
+		User:            cfg.User,
+		Image:           cfg.Image,
+		NetworkDisabled: cfg.DisableNetwork,
+		Labels:          cfg.Labels,
+	}
+	containerConfig.Env = append(containerConfig.Env, fmt.Sprintf("%s=%s", plugin.EnvUnixSocketDir, pluginSocketDir))
+	if cmd.Dir != "" {
+		containerConfig.WorkingDir = cmd.Dir
+	}
+	if cmd.Path != "" {
+		containerConfig.Entrypoint = []string{cmd.Path}
+	}
+	if cmd.Args != nil {
+		containerConfig.Cmd = cmd.Args
+		containerConfig.ArgsEscaped = true
+	}
+
+	// Host config.
+	// TODO: Can we safely we drop some default capabilities?
+	hostConfig := &container.HostConfig{
+		AutoRemove:    true,                      // Plugin containers are ephemeral.
```

Comment:
Is this behavior possible due to Vault's lazy-loading? i.e. if the container is not running, Vault will restart it when the plugin receives a request.

Reply:
I'm not sure I quite understand the question. The reasoning for this setting is to reduce management overhead - we _could_ keep restarting the same container, but then we need to be more careful about tracking when we can delete it. Whereas if we just always delete containers after the plugin process has exited, then it mirrors the lifecycle of a plain plugin process more closely.

Reply:
It's equivalent to the `--rm` setting for `docker run` btw

Reply:
That makes sense, thanks!